src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+where
+    T: ArrowPrimitiveType,
+{
+    let mut decoded_results = Vec::with_capacity(encoded_results.len());
+    for i in encoded_results {
+        decoded_results.push(Some(decoder(i)?));
+    }
+    Ok(decoded_results.iter().collect::<PrimitiveArray<T>>())
+}
+
+fn messagepack_decode_results(
+    return_type: &CreateFunctionDataType,
+    encoded_results: &Vec<Value>,
+) -> Result<ArrayRef> {
+    match return_type {
+        CreateFunctionDataType::SMALLINT => decode_udf_result_primitive_array::<
+            arrow::datatypes::Int16Type,
+        >(encoded_results, &|v| match v
+            .as_i64()
+            .ok_or(DataFusionError::Internal(format!(
+                "Expected to find i64 value, but received {:?} instead",
+                v
+            ))) {
+            Err(e) => Err(e),
+            Ok(v_i64) => i16::try_from(v_i64).map_err(|e| {
+                DataFusionError::Internal(format!("Error converting i64 to i16: {:?}", e))
+            }),
+        })
+        .map(|a| Arc::new(a) as ArrayRef),
+        CreateFunctionDataType::I32 | CreateFunctionDataType::INT => {
+            decode_udf_result_primitive_array::<arrow::datatypes::Int32Type>(
+                encoded_results,
+                &|v| match v.as_i64().ok_or(DataFusionError::Internal(format!(
+                    "Expected to find i64 value, but received {:?} instead",
+                    v
+                ))) {
+                    Err(e) => Err(e),
+                    Ok(v_i64) => i32::try_from(v_i64).map_err(|e| {
+                        DataFusionError::Internal(format!(
+                            "Error converting i64 to i32: {:?}",
+                            e
+                        ))
+                    }),
+                },
+            )
+            .map(|a| Arc::new(a) as ArrayRef)
+        }
+        CreateFunctionDataType::I64 | CreateFunctionDataType::BIGINT => {
+            decode_udf_result_primitive_array::<arrow::datatypes::Int64Type>(
+                encoded_results,
+                &|v| {
+                    v.as_i64().ok_or(DataFusionError::Internal(format!(
+                        "Expected to find i64 value, but received {:?} instead",
+                        v
+                    )))
+                },
+            )
+            .map(|a| Arc::new(a) as ArrayRef)
+        }
+        CreateFunctionDataType::CHAR
+        | CreateFunctionDataType::VARCHAR
+        | CreateFunctionDataType::TEXT => {
+            let mut decoded_results = Vec::with_capacity(encoded_results.len());
+            for i in encoded_results {
+                decoded_results.push(Some(i.as_str().ok_or(
+                    DataFusionError::Internal(format!(
+                        "Expected to find string value, received {:?} instead",
+                        &i
+                    )),
+                )?));
+            }
+            Ok(Arc::new(decoded_results.iter().collect::<StringArray>()) as ArrayRef)
+        }
```

Comment:
Does `decode_udf_result_primitive_array` not work here because it's not a `PrimitiveArray`? You could also rewrite this function to avoid mut + for:

```suggestion
            let decoded_results = encoded_results.iter().map(|i| (i.as_str().ok_or(
                DataFusionError::Internal(format!(
                    "Expected to find string value, received {:?} instead",
                    &i
                )),
            ))).collect::<Result<StringArray>>()?;
            Ok(Arc::new(decoded_results) as ArrayRef)
        }
```